File: python/constant_medium.py
import sys
import math
import random

from hitable import *
from material import *
from util import *

class Isotropic(Material):
	def __init__(self, albedo):
		self.albedo = albedo

	def scatter(self, ray_in, hit_record):
		target = hit_record.pt + hit_record.normal + pt_in_unit_sphere()
		scattered = Ray(hit_record.pt, pt_in_unit_sphere(), ray_in.time)
		return (True, scattered, self.albedo.value(hit_record.u, hit_record.v, hit_record.pt))

class ConstantMedium(Hitable):
	def __init__(self, boundary, density, phaseTexture):
		self.boundary = boundary
		self.density = density
		self.material = Isotropic(phaseTexture)

	def hit(self, ray, tmin, tmax):
		hit1, rec1 = self.boundary.hit(ray, -sys.float_info.max, sys.float_info.max)
		if hit1:
			hit2, rec2 = self.boundary.hit(ray, rec1.t + 0.0001, sys.float_info.max)
			if hit2:
				rec1.t = max(rec1.t, tmin)
				rec2.t = min(rec2.t, tmax)
				if rec1.t < 0:
					return (False, None)
				distInsideBoundary = (rec2.t - rec1.t) * ray.direction.length()
				hitDistance = -(1.0 / self.density)*math.log(random.random())
				if hitDistance < distInsideBoundary:
					t = rec1.t + hitDistance / ray.direction.length()
					pt = ray.point_at_t(t)
					return (True, HitRecord(t, pt, Vector3(0,1,0), 0, 0, self.material))
		return (False, None)

	def bounding_box(self, sceneTime0, sceneTime1):
		return self.boundary.bounding_box(sceneTime0, sceneTime1)

File: rust/raytrace/src/constant_medium.rs
use std::sync::Arc;

use hitable::{Hitable, HitablePtr, HitRecord};
use material::{Material, MaterialPtr, ScatterInfo};
use texture::TexturePtr;
use rt_rand::*;
use ray::Ray;
use aabb::Aabb;
use vec3::Vec3;

pub struct Isotropic {
  albedo: TexturePtr,
}

impl Isotropic {
  pub fn new(albedo: TexturePtr) -> Isotropic {
    Isotropic {
      albedo: Arc::clone(&albedo)
    }
  }

  pub fn rc(albedo: TexturePtr) -> Arc<Isotropic> {
    Arc::new(Isotropic::new(albedo))
  }
}

impl Material for Isotropic {
  fn scatter(&self, ray: &Ray, hit: &HitRecord) -> Option<ScatterInfo> {
    let scattered = Ray::new(hit.p, random_in_unit_sphere(), ray.time);
    Some(ScatterInfo {
      attenuation: self.albedo.value(hit.u, hit.v, &hit.p),
      scattered
    })
  }
}

pub struct ConstantMedium {
  boundary: HitablePtr,
  density: f64,
  material: MaterialPtr
}

impl ConstantMedium {
  pub fn new(boundary: &HitablePtr, density: f64, phase_texture: TexturePtr) -> ConstantMedium {
    let material: MaterialPtr = Isotropic::rc(phase_texture);
    ConstantMedium {
      boundary: Arc::clone(boundary),
      density,
      material
    }
  }

  pub fn hitable_ptr(boundary: &HitablePtr, density: f64, phase_texture: TexturePtr) -> Arc<ConstantMedium> {
    Arc::new(ConstantMedium::new(boundary, density, phase_texture))
  }
}

impl Hitable for ConstantMedium {
    fn hit(&self, ray: &Ray, t_min: f64, t_max: f64) -> Option<HitRecord> {
      if let Some(mut hit1) = self.boundary.hit(ray, -std::f64::MAX, std::f64::MAX) {
        if let Some(mut hit2) = self.boundary.hit(ray, hit1.t + 0.0001, std::f64::MAX) {
          hit1.t = hit1.t.max(t_min);
          hit2.t = hit2.t.min(t_max);
          if hit1.t < 0.0 {
            return None;
          }
          let dist_inside_boundary = (hit2.t - hit1.t) * ray.direction.length();
          let hit_distance = -(1.0 / self.density) * rand_f64().log(std::f64::consts::E);
          if hit_distance < dist_inside_boundary {
            let t = hit1.t + hit_distance / ray.direction.length();
            let pt = ray.point_at_parameter(t);
            return Some(HitRecord::new(t, pt, Vec3::new(0.0, 0.0, 0.0), 0.0, 0.0, self.material.clone()));

          }
        }
      }
      None
    }

    fn bounding_box(&self, time0: f64, time1: f64) -> Aabb {
      self.boundary.bounding_box(time0, time1)
    }
}